src/lib/storage/index/partial_hash/partial_hash_index_impl.cpp

```diff
@@ -0,0 +1,253 @@
+  };
+
+  // Iterate over all values stored in the index.
+  auto map_iter = _map.begin();
+  while (map_iter != _map.end()) {
+    auto& row_ids = _map.at(map_iter->first);
+
+    // Remove every RowID entry of the value that references one of the chunks.
+    row_ids.erase(std::remove_if(row_ids.begin(), row_ids.end(), is_to_unindex), row_ids.end());
+
+    map_iter = row_ids.empty() ? _map.erase(map_iter) : ++map_iter;
+  }
+
+  auto& nulls = _null_values;
+
+  nulls.erase(std::remove_if(nulls.begin(), nulls.end(), is_to_unindex), nulls.end());
+
+  return size_before - _indexed_chunk_ids.size();
+}
+
+template <typename DataType>
+typename PartialHashIndexImpl<DataType>::IteratorPair PartialHashIndexImpl<DataType>::range_equals(
+    const AllTypeVariant& value) const {
+  const auto begin = _map.find(boost::get<DataType>(value));
+  if (begin == _map.end()) {
+    const auto end_iter = cend();
+    return std::make_pair(end_iter, end_iter);
+  }
+  auto end = begin;
+  return std::make_pair(Iterator(std::make_shared<TableIndexFlattenedSparseMapIterator<DataType>>(begin)),
+                        Iterator(std::make_shared<TableIndexFlattenedSparseMapIterator<DataType>>(++end)));
+}
+
+template <typename DataType>
+std::pair<typename PartialHashIndexImpl<DataType>::IteratorPair, typename PartialHashIndexImpl<DataType>::IteratorPair>
+PartialHashIndexImpl<DataType>::range_not_equals(const AllTypeVariant& value) const {
+  const auto range_eq = range_equals(value);
+  return std::make_pair(std::make_pair(cbegin(), range_eq.first), std::make_pair(range_eq.second, cend()));
+}
+
+template <typename DataType>
+typename PartialHashIndexImpl<DataType>::Iterator PartialHashIndexImpl<DataType>::cbegin() const {
+  return Iterator(std::make_shared<TableIndexFlattenedSparseMapIterator<DataType>>(_map.cbegin()));
+}
+
+template <typename DataType>
+typename PartialHashIndexImpl<DataType>::Iterator PartialHashIndexImpl<DataType>::cend() const {
+  return Iterator(std::make_shared<TableIndexFlattenedSparseMapIterator<DataType>>(_map.cend()));
+}
+
+template <typename DataType>
+typename PartialHashIndexImpl<DataType>::Iterator PartialHashIndexImpl<DataType>::null_cbegin() const {
+  return Iterator(std::make_shared<TableIndexVectorIterator>(_null_values.cbegin()));
+}
+
+template <typename DataType>
+typename PartialHashIndexImpl<DataType>::Iterator PartialHashIndexImpl<DataType>::null_cend() const {
+  return Iterator(std::make_shared<TableIndexVectorIterator>(_null_values.cend()));
+}
+
+template <typename DataType>
+size_t PartialHashIndexImpl<DataType>::memory_usage() const {
+  auto bytes = size_t{0u};
+
+  bytes += sizeof(_indexed_chunk_ids);  // NOLINT - Linter complains I should use vector.size(), which would be wrong.
+  bytes += sizeof(ChunkID) * _indexed_chunk_ids.size();
+
+  bytes += sizeof(_map);
+  // TODO(anyone): Find a clever way to estimate the hash sizes in the maps. For now we estimate a hash size of 8 byte
```

Comment:
Does the sparse map change the type of the hashes returned by the provided hash function? If not, then the hash size is equal to the size of the type returned by the hash function (default: `std::hash` &rarr; `size_t`).

Reply:
Good to know, thanks!